### PR TITLE
Migrate remote db package to kotlin

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
Towards "Alpha cleanup" milestone

Migrates all remote firestore classes to kotlin:
* `data` for models classes
* `object` for static classes
* `class` for the remaining cases

Tested by creating a fresh project using web console and opening it in the app.

@gino-m @scolsen @JSunde PTAL?

P.S. The kotlin migration is done using AS's migrate tool. So it would be easier to review the code commit-by-commit

### CODE REVIEW

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/base/FluentCollectionReference.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import java8.util.function.Function
+
+abstract class FluentCollectionReference protected constructor(private val reference: CollectionReference) {
+
+    /**
+     * Returns a Completable that completes immediately on subscribe if network is available, or fails
+     * in error if not.
+     */
+    private fun requireActiveNetwork(): @Cold Completable {
+        return requireActiveNetwork(
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/base/FluentCollectionReference.kt:36]: nit: could be a single line expression function

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/base/FluentCollectionReference.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        return requireActiveNetwork()
+            .andThen(
+                FluentFirestore.toSingleList(
+                    RxFirestore.getCollection(query),
+                    mappingFunction
+                )
+            )
+    }
+
+    protected fun reference(): CollectionReference {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/base/FluentCollectionReference.kt:58]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/base/FluentCollectionReference.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        return reference
+    }
+
+    override fun toString(): String {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/base/FluentCollectionReference.kt:62]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/base/FluentDocumentReference.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    protected fun merge(values: ImmutableMap<String, Any>, batch: WriteBatch) {
+        batch[reference, values] = SetOptions.merge()
+    }
+
+    /** Adds a request to the specified batch to delete the current DocumentReference.  */
+    protected fun delete(batch: WriteBatch) {
+        batch.delete(reference)
+    }
+
+    protected fun reference(): DocumentReference {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/base/FluentDocumentReference.kt:39]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/base/FluentDocumentReference.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        return reference
+    }
+
+    override fun toString(): String {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/base/FluentDocumentReference.kt:43]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/base/FluentFirestore.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import com.google.firebase.firestore.QuerySnapshot
+import com.google.firebase.firestore.WriteBatch
+import io.reactivex.Maybe
+import io.reactivex.Single
+import java8.util.function.Function
+
+/** Base class for representing Firestore databases as object hierarchies.  */
+abstract class FluentFirestore protected constructor(private val db: FirebaseFirestore) {
+
+    protected fun db(): FirebaseFirestore {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/base/FluentFirestore.kt:31]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/AuditInfoConverter.kt [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        DataStoreException.checkNotNull(doc.clientTimestamp, "clientTimestamp")
+        return AuditInfo.builder()
+            .setUser(UserConverter.toUser(doc.user))
+            .setClientTimestamp(doc.clientTimestamp!!.toDate())
+            .setServerTimestamp(Optional.ofNullable(doc.serverTimestamp?.toDate()))
+            .build()
+    }
+
+    @JvmStatic
+    fun fromMutationAndUser(mutation: Mutation, user: User): AuditInfoNestedObject {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/AuditInfoConverter.kt:40]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiCollectionReference.kt [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+class LoiCollectionReference internal constructor(ref: CollectionReference) :
+    FluentCollectionReference(ref) {
+
+    /** Retrieves all lois in the survey, then streams changes to the remote db incrementally.  */
+    fun loadOnceAndStreamChanges(survey: Survey): @Cold(terminates = false) Flowable<RemoteDataEvent<LocationOfInterest?>> {
+        return RxFirestore.observeQueryRef(reference())
+            .flatMapIterable { snapshot: QuerySnapshot -> toRemoteDataEvents(survey, snapshot) }
+    }
+
+    fun loi(id: String): LoiDocumentReference {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiCollectionReference.kt:40]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiCollectionReference.kt [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import com.google.android.ground.persistence.remote.firestore.base.FluentCollectionReference
+import com.google.android.ground.persistence.remote.firestore.schema.LoiConverter.toLoi
+import com.google.android.ground.rx.annotations.Cold
+import com.google.firebase.firestore.CollectionReference
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.QuerySnapshot
+import durdinapps.rxfirebase2.RxFirestore
+import io.reactivex.Flowable
+
+class LoiCollectionReference internal constructor(ref: CollectionReference) :
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiCollectionReference.kt:31]: Can you add some kdoc?

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiCollectionReference.kt [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    fun loi(id: String): LoiDocumentReference {
+        return LoiDocumentReference(reference().document(id))
+    }
+
+    private fun toRemoteDataEvents(
+        survey: Survey,
+        snapshot: QuerySnapshot
+    ): Iterable<RemoteDataEvent<LocationOfInterest?>> {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiCollectionReference.kt:47]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiDocumentReference.kt [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+package com.google.android.ground.persistence.remote.firestore.schema
+
+import com.google.android.ground.model.User
+import com.google.android.ground.model.mutation.LocationOfInterestMutation
+import com.google.android.ground.model.mutation.Mutation
+import com.google.android.ground.persistence.remote.firestore.base.FluentDocumentReference
+import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.WriteBatch
+
+class LoiDocumentReference internal constructor(ref: DocumentReference) :
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiDocumentReference.kt:25]: Can you add kdoc?

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiMutationConverter.kt [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+            }
+            Mutation.Type.UPDATE ->
+                map.put(LoiConverter.LAST_MODIFIED, auditInfo)
+            Mutation.Type.DELETE, Mutation.Type.UNKNOWN ->
+                throw UnsupportedOperationException()
+        }
+        return map.build()
+    }
+
+    private fun toGeoPoint(point: Point): GeoPoint {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiMutationConverter.kt:68]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiMutationConverter.kt [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        return GeoPoint(point.latitude, point.longitude)
+    }
+
+    private fun toGeoPointList(point: ImmutableList<Point>): List<GeoPoint> {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiMutationConverter.kt:72]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/MultipleChoiceConverter.kt [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+ * limitations under the License.
+ */
+
+package com.google.android.ground.persistence.remote.firestore.schema
+
+import com.google.android.ground.model.task.MultipleChoice
+import com.google.android.ground.util.Enums.toEnum
+import com.google.android.ground.util.toImmutableList
+
+internal object MultipleChoiceConverter {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/MultipleChoiceConverter.kt:23]: kdoc

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionCollectionReference.kt [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+import com.google.android.ground.util.toImmutableList
+import com.google.common.collect.ImmutableList
+import com.google.firebase.firestore.*
+import durdinapps.rxfirebase2.RxFirestore
+import io.reactivex.Single
+
+class SubmissionCollectionReference internal constructor(ref: CollectionReference) :
+    FluentCollectionReference(ref) {
+
+    fun submission(id: String): SubmissionDocumentReference {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionCollectionReference.kt:34]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionCollectionReference.kt [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        return querySnapshot.documents
+            .map { doc: DocumentSnapshot ->
+                create {
+                    SubmissionConverter.toSubmission(locationOfInterest, doc)
+                }
+            }
+            .toImmutableList()
+    }
+
+    private fun byLoiId(loiId: String): Query {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionCollectionReference.kt:58]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionConverter.kt [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+        val value = DataStoreException.checkType(Timestamp::class.java, obj) as Timestamp
+        TimeResponse.fromDate(value.toDate())
+            .ifPresent { r: Response -> responses.putResponse(taskId, r) }
+    }
+
+    private fun putMultipleChoiceResponse(
+        taskId: String, multipleChoice: MultipleChoice?, obj: Any, responses: ResponseMap.Builder
+    ) {
+        val values = DataStoreException.checkType(MutableList::class.java, obj) as List<*>
+        values.forEach { v: Any? ->
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionConverter.kt:124]: Could the "v: Any?" be removed and just use "it" instead?

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionConverter.kt [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+            DataStoreException.checkType(String::class.java, v as Any)
+        }
+        MultipleChoiceResponse.fromList(multipleChoice, values as List<String>)
+            .ifPresent { r: Response -> responses.putResponse(taskId, r) }
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionConverter.kt:128]: "r: Response" could probably be omitted as well, unless you think it's needed for readability

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionDocumentReference.kt [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+import com.google.android.ground.model.submission.Submission
+import com.google.android.ground.persistence.remote.firestore.base.FluentDocumentReference
+import com.google.android.ground.rx.annotations.Cold
+import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.WriteBatch
+import durdinapps.rxfirebase2.RxFirestore
+import io.reactivex.Maybe
+
+class SubmissionDocumentReference internal constructor(ref: DocumentReference) :
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionDocumentReference.kt:32]: kdoc

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SurveyDocumentReference.kt [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+import durdinapps.rxfirebase2.RxFirestore
+import io.reactivex.Maybe
+
+private const val LOIS = "lois"
+private const val SUBMISSIONS = "submissions"
+
+class SurveyDocumentReference internal constructor(ref: DocumentReference) :
+    FluentDocumentReference(ref) {
+
+    fun lois(): LoiCollectionReference {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SurveyDocumentReference.kt:32]: single line expression function

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SurveyDocumentReference.kt [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+        return LoiCollectionReference(reference().collection(LOIS))
+    }
+
+    fun submissions(): SubmissionCollectionReference {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SurveyDocumentReference.kt:36]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SurveyDocumentReference.kt [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+        return SubmissionCollectionReference(reference().collection(SUBMISSIONS))
+    }
+
+    fun get(): Maybe<Survey> {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SurveyDocumentReference.kt:40]: ditto

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SurveysCollectionReference.kt [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+import io.reactivex.Single
+
+private const val ACL_FIELD = "acl"
+private const val OWNER_ROLE = "owner"
+private const val MANAGER_ROLE = "manager"
+private const val CONTRIBUTOR_ROLE = "contributor"
+private const val VIEWER_ROLE = "viewer"
+private val VALID_ROLES = listOf(OWNER_ROLE, MANAGER_ROLE, CONTRIBUTOR_ROLE, VIEWER_ROLE)
+
+class SurveysCollectionReference internal constructor(ref: CollectionReference) :
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SurveysCollectionReference.kt:35]: kdoc

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SurveysCollectionReference.kt [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+    FluentCollectionReference(ref) {
+
+    fun survey(id: String): SurveyDocumentReference {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SurveysCollectionReference.kt:38]: single line expression function

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/UserConverter.kt [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ */
+
+package com.google.android.ground.persistence.remote.firestore.schema
+
+import com.google.android.ground.model.User
+
+/** Converts between Firestore objects and [User] instances.  */
+internal object UserConverter {
+    fun toNestedObject(user: User): UserNestedObject {
```
Comment (JSunde) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/UserConverter.kt:23]: single line expression function

**Review by JSunde — APPROVED** [review]

**Comment by shobhitagarwal1612** [comment]
PTAL. Tested locally.

**Pull request merged**